### ISSUE TITLE
`@distributed` fails on loops over iterators

### ISSUE THREAD

**Issue by dangirsh** [JuliaLang/julia#30343]
If you have a question or are unsure if the behavior you're experiencing is a bug,
please search or post to our Discourse site: https://discourse.julialang.org. We use
the GitHub issue tracker for bug reports and feature requests only.

If you're submitting a bug report, be sure to include as much relevant information as
possible, including a minimal reproducible example and the output of `versioninfo()`.
If you're experiencing a problem with a particular package, open an issue on that
package's repository instead.

Thanks for contributing to the Julia project!

**Comment by sunnsi** [comment]
Does any one tell me whether the  macro "@distributed" work on iterators now? I have a very large data array (more than 900,000,000)  after combinations,  collect() function will exhaust all memory.

**Comment by mbauman** [comment]
Yes, `@distributed` still requires indexing — it'd indeed be good to document this requirement.

Note that distributing such a large array to your workers may swamp the computation time with communication overhead. You may want to consider "peeling off" one of your combinatorial elements, distribute that amongst your workers, and then have each of your workers compute the combinations of that one element with the remainder.

**Comment by sunnsi** [comment]
Thanks, mbauman. I will try your suggestions.

**Comment by bocc** [comment]
Hello Matt,

if I understand correctly, this indexing requirement is the reason behind `eachrow` & `eachcol` not working with `@distributed`? Julia reports a syntax error saying 'invalid assignment location' in 1.2.

**Comment by mbauman** [comment]
Yes, that's correct.  Just as a breadcrumb, with https://github.com/JuliaLang/julia/pull/32310 we could possibly add indexing support.
